pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -519,7 +519,7 @@ func (ctrl *PersistentVolumeController) syncVolume(volume *v1.PersistentVolume)
 		if err != nil {
 			return err
 		}
-		if !found && metav1.HasAnnotation(volume.ObjectMeta, pvutil.AnnBoundByController) {
+		if !found {
```

Comment:
hi, you should add `pvutil.AnnBoundByController` annotation in your external PV binder. With this annotation, PV controller will go into this block and check apiserver if not found in informer cache. This is required to work well with Kubernetes PV controller: https://github.com/kubernetes/kubernetes/blob/32ffe47fd2d124096f42d843f29d6b123b901d52/pkg/controller/volume/persistentvolume/util/util.go#L48-L50.

This annotation is used to distinguish from PVs & PVCs bound manually, e.g. create PVC with PV bound.

Reply:
Actually, I does not use any external PV binder, I use CSI provisioner to create PV then meet this bug. PVs that provisioned by CSI provisioner does not have this annotation:

https://github.com/kubernetes-csi/external-provisioner/blob/e17cc20ee72c6c0bb27a3cf81348f40834925136/vendor/sigs.k8s.io/sig-storage-lib-external-provisioner/controller/controller.go#L1280-L1281

And for now, there is no way to develop our  external PV binder, because the pv-binder-controller will take over all PVC/PVs and there is no way to skip pv-binder-controller management to use our external PV binder.

Reply:
oh, I see the problem. External provisioner created volumes with `ClaimRef` filled and wait for the PV controller to bind them with. However, PV controller will reclaim these PVs if they don't see the related PVCs in `ctrl.claims`. (PV is created before PVC is put into `ctrl.claims`?)

I think, remove `metav1.HasAnnotation(volume.ObjectMeta, pvutil.AnnBoundByController)` is enough here because it will check informer cache, no need to update `ctrl.claims` in informer listeners (informer cache is updated first).

Alternatively, we can ignore PVs not bound by PV controller and don't reclaim them [here](https://github.com/kubernetes/kubernetes/blob/654aa1d846a72ba842967424d287304db1dad803/pkg/controller/volume/persistentvolume/pv_controller.go#L565-L588).

cc @msau42 @jsafrane What's your opinions?

Reply:
> PV is created by external provisioner after dynamic provisioning is triggered and before PVC is put into ctrl.claims?

Yes, this is the bug root case.

> no need to update ctrl.claims in informer listeners

I think the reason why the controller need `ctrl.claims` is that the controller want to get the latest version of PVC/PV, update the cache when controller finish updating PVC/PV is only guaranteed when update by pv-bind-controller side. But on the other side, any other controller may also update PVC/PVs and pv-bind-controller receive the update event, in that case we need update `ctrl.claims` also.

> we can ignore PVs not bound by PV controller and don't reclaim them

Actually, I'm draft a KEP to make possible to develop a high performance component to  manage PVC/PV lifecycle, typically for CSI volumes. And for now, there is no way to develop our external PV binder, and the pv-bind-controller performance does not meet our requirement, performance issue see #83178.